### ISSUE TITLE
Various issues with generated Swagger docs

### ISSUE THREAD

**Issue by Jus-Codin** [CTFd/CTFd#2632]
**Environment**:

- CTFd Version/Commit: https://github.com/CTFd/CTFd/commit/13efb55e73f378d4a3b88f07ed983848b4ab2458

## TL;DR
There are several issues with the swagger.json file generated by CTFd, and most of them is due to the [`validate_args`](https://github.com/CTFd/CTFd/blob/13efb55e73f378d4a3b88f07ed983848b4ab2458/CTFd/api/v1/helpers/request.py#L15) decorator.

Issues are as follows:
1. Missing `$ref` type definitions (as mentioned in #2574)
2. Incorrect specification of parameters for POST JSON body endpoints
3. Duplicate `get_flag_types` operationId for `/flag/types` and `/flag/types/<type_name>` endpoint

I'm interested in fixing these issues, but would like some feedback to see if the fixes are going in the right direction.

## Missing `$ref` type definitions (refer to above issue for details)
This issue is caused by the `RawEnum` definitions for the `field` query parameters. Pydantic converts these to references in the model schema to reduce repeated enum definitions, but `validate_args` [only includes `properties` from the schema](https://github.com/CTFd/CTFd/blob/13efb55e73f378d4a3b88f07ed983848b4ab2458/CTFd/api/v1/helpers/request.py#L38), resulting in the missing definition.

<details>
<summary>Sample affected schema (AwardFields)</summary>

```json
{
    "title": "",
    "type": "object",
    "properties": {
        "user_id": {
            "title": "User Id",
            "type": "integer"
        },
        ...
        "field": {
            "$ref": "#/definitions/AwardFields"
        }
    },
    "definitions": {
        "AwardFields": {
            "title": "AwardFields",
            "description": "This is a customized enum class which should be used with a mixin.\nThe mixin should define the types of each member.\n\nFor example:\n\nclass Colors(str, RawEnum):\n    RED = \"red\"\n    GREEN = \"green\"\n    BLUE = \"blue\"",
            "enum": [
                "name",
                "description",
                "category",
                "icon"
            ]
        }
    }
}
```
</details>

I have created a fix on my [fork](https://github.com/Jus-Codin/CTFd/blob/testing/CTFd/api/v1/helpers/request.py#L36-L39), which just resolves the reference and adds in the `enum` field.

## Incorrect specification of parameters for POST JSON body endpoints
According to https://swagger.io/docs/specification/v2_0/describing-parameters/#parameter-types, `json` is not a valid value for the `in` field in parameters.

Currently, the parameters for a JSON POST body looks like this:

```yaml
parameters:
    - type: string
      in: json
      name: example
```

When it should look like:

```yaml
parameters:
    - in: body
      name: page
      description: The page to create.
      schema:
        type: object
        required:
          - example
        properties:
          example:
            type: string
```

Additionally, `title` should be removed all parameter fields as it isn't part of the Swagger specification.

I suggest fixing this by checking if `location="json"` in `validate_args` and correctly formatting the dict.

## Duplicate `get_flag_types` operationId for `/flag/types` and `/flag/types/<type_name>` endpoint
The [`FlagTypes`](https://github.com/CTFd/CTFd/blob/master/CTFd/api/v1/flags.py#L108-L125) resource is used by `/flag/types` and `/flag/types/<type_name>`. This has the unintended effect of making these routes both have the same `operationId` which is not allowed:

```json
"/flags/types": {
    "get": {
        "responses": {
            "200": {
                "description": "Success"
            }
        },
        "operationId": "get_flag_types",
        "tags": [
            "flags"
        ]
    }
},
"/flags/types/{type_name}": {
    "parameters": [
        {
            "name": "type_name",
            "in": "path",
            "required": true,
            "type": "string"
        }
    ],
    "get": {
        "responses": {
            "200": {
                "description": "Success"
            }
        },
        "operationId": "get_flag_types",
        "tags": [
            "flags"
        ]
    }
}
```

Suggested fix is to separate the endpoints as shown [here](https://github.com/Jus-Codin/CTFd/blob/testing/CTFd/api/v1/flags.py#L108-L128).

If these issues are resolved, it should be possible to automatically generate API clients for CTFd.

**Comment by ColdHeat** [comment]
You could PR the changes that you're suggesting here.

**Comment by ColdHeat** [comment]
Now that #2633 has been merged, could you maybe make new issues for the remaining items? Or you could edit this one.

**Comment by Jus-Codin** [comment]
Thanks for the merge, since most of the problems in this issue is fixed, I think it would be better to work on other changes in a new issue.
